cmd/elementGen/elements.go

```diff
@@ -0,0 +1,727 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"go/format"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"text/template"
+)
+
+type (
+	// Desc contains a description of each HTML element that should be generated.
+	// - Name is the Pascal-case version (e.g. BaseFont).
+	// - JS is the lowercase version (e.g. basefont).
+	// - Override is used if the JS version is not the lowercase of the name (e.g. Script -> s).
+	// - Basic indicates that in "honnef.co/go/js/dom", we should use dom.BasicHTMLElement instead of
+	//   HTMLXXXElement, where XXX is the element name (e.g. dom.HTMLAppletElement).
+	// - DOM is the selected "honnef.co/go/js/dom" name (e.g. either dom.BasicHTMLElement or
+	//   dom.HTMLAppletElement).
+	Desc struct {
+		Name       string
+		JS         string
+		Override   string
+		Attributes []Attr
+		Basic      bool // True if honnef.co/go/js/dom declares a specific Element type.
+		DOM        string
+	}
+
+	// Attr contains information about each non-standard HTML attribute that an element supports.
+	// - Name is the Pascal-case version (e.g. DataSrc).
+	// - JS is the lowercase version (e.g. datasrc).
+	// - Type is the Go type (e.g. bool) if it is not "string".
+	Attr struct {
+		Name string
+		JS   string
+		Type string
+	}
+)
+
+func generate() {
+	primary := template.Must(template.New("primary").Parse(primaryTemplate))
+	test := template.Must(template.New("test").Parse(testTemplate))
+
+	var names []string
+	for k := range elements {
+		names = append(names, k)
+	}
+	sort.Strings(names)
+
+	var ele []Desc
+
+	for _, n := range names {
+		v := elements[n]
+		v.Name = n
+		v.JS = strings.ToLower(n)
+		if v.Override != "" {
+			v.Name = v.Override
+		}
+
+		v.DOM = fmt.Sprintf("HTML%sElement", v.Name)
+		if v.Basic {
+			v.DOM = "BasicHTMLElement"
+		}
+
+		for i := 0; i < len(v.Attributes); i++ {
+			a := &v.Attributes[i]
+			a.JS = strings.ToLower(a.Name)
+			if a.Type == "" {
+				a.Type = "string"
+			}
+		}
+
+		ele = append(ele, v)
+	}
+
+	executeTemplate("gen_elements_elementGen.go", primary, ele)
+	executeTemplate("gen_elements_elementGen_test.go", test, ele)
+}
+
+func executeTemplate(filename string, t *template.Template, ele []Desc) {
+	var b bytes.Buffer
+
+	if err := t.Execute(&b, ele); err != nil {
+		panic(err)
```

Comment:
I'd prefer to do [something like this](https://github.com/myitcv/react/blob/f32f0c5d7468766a687ac82cef960f5f64b72270/cmd/reactGen/react_gen.go#L85-L87) for now. Neither is particularly ideal, but at least with `fatalf` (or equivalent) we can control whether we want to panic/do something else. It also means we can then easily have `infof` in case we want to log verbose messages etc.